torchtune/models/flux/_autoencoder.py

```diff
@@ -0,0 +1,352 @@
+
+        self.conv_in = nn.Conv2d(ch_in, channels[0], kernel_size=3, stride=1, padding=1)
+
+        self.down = nn.Sequential(
+            *[
+                DownBlock(
+                    n_layers=n_layers_per_down_block,
+                    ch_in=channels[i - 1] if i > 0 else channels[0],
+                    ch_out=channels[i],
+                    downsample=i < len(channels) - 1,
+                )
+                for i in range(len(channels))
+            ]
+        )
+
+        self.mid = mid_block(channels[-1])
+
+        self.end = end_block(channels[-1], 2 * ch_z)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """
+        Args:
+            x (Tensor): input images (shape = [bsz, ch_in, img resolution, img resolution])
+
+        Returns:
+            Tensor: latent encodings (shape = [bsz, ch_z, latent resolution, latent resolution])
+        """
+        h = self.conv_in(x)
+        h = self.down(h)
+        h = self.mid(h)
+        h = self.end(h)
+        z = diagonal_gaussian(h)
+        return self.scale_factor * (z - self.shift_factor)
+
+
+class Decoder(nn.Module):
+    """
+    The encoder half of the Flux diffusion model's image autoencoder.
+
+    Args:
+        ch_out (int): The number of channels of the output image.
+        ch_z (int): The number of latent channels (dimension of the latent vector `z`).
+        channels (List[int]): The number of output channels for each upsample block.
+        n_layers_per_up_block (int): Number of resnet layers per upsample block.
+        scale_factor (float): Constant for scaling `z`.
+        shift_factor (float): Constant for shifting `z`.
+    """
+
+    def __init__(
+        self,
+        ch_out: int,
+        ch_z: int,
+        channels: List[int],
+        n_layers_per_up_block: int,
+        scale_factor: float,
+        shift_factor: float,
+    ):
+        super().__init__()
+        self.scale_factor = scale_factor
+        self.shift_factor = shift_factor
+
+        self.conv_in = nn.Conv2d(ch_z, channels[0], kernel_size=3, stride=1, padding=1)
+
+        self.mid = mid_block(channels[0])
+
+        self.up = nn.Sequential(
+            *[
+                UpBlock(
+                    n_layers=n_layers_per_up_block,
+                    ch_in=channels[i - 1] if i > 0 else channels[0],
+                    ch_out=channels[i],
+                    upsample=i < len(channels) - 1,
+                )
+                for i in range(len(channels))
+            ]
+        )
+
+        self.end = end_block(channels[-1], ch_out)
+
+    def forward(self, z: Tensor) -> Tensor:
+        """
+        Args:
+            z (Tensor): latent encodings (shape = [bsz, ch_z, latent resolution, latent resolution])
+
+        Returns:
+            Tensor: output images (shape = [bsz, ch_in, img resolution, img resolution])
+        """
+        z = z / self.scale_factor + self.shift_factor
+        h = self.conv_in(z)
+        h = self.mid(h)
```

Comment:
Small thing, we follow the pattern of looping through the list of layers here instead of putting them into Sequential. This makes the forward pass a bit more readable and easier to add breakpoints.